Port check_grep.sh from an increasingly complicated shell script to Python.

warriorstar-orion opened the pull request:
🚨 _It is expected that this PR fails CI until all findings are fixed in the codebase._ 🚨

## What Does This PR Do

This PR is a straightforward port of check_grep.sh (including additional checks from #19837 that were not submitted) to Python. It is intentionally as rote as possible: no caching, no abstraction, no fancy test suites or error collectors. This is to have as few moving parts when verifying the tests have been ported properly, and make it easier to reason about when adding new tests/figuring out where a test should go (in a linter or an in-engine map test, etc).

## Why It's Good For The Game
There comes a time in every shell script's life when it has to evolve into something more manageable and stateful. That time for check_grep has come. I believe this will come with performance benefits as well, due to some built-in Python interpreter caching, as well as being able to cut out of a test early on the first line it finds an issue on for each file.

And it could probably be an order of magnitude faster if we iterate over files -> tests, rather than tests -> files as it currently is for parity with the existing test structure.

Regexes for indentation (whitespace and tabs) shamelessly stolen from StackOverflow.

## Testing

Testing still in progress, but run locally, and spot checking results, they seem to be accurate, but I'll need to come up with more test cases to be certain.

Pretty obscure infra change so no CL block here.

lewcc commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,337 @@
+def check_comments(filename, lines):
+    for line in lines:
+        if (
+            re.match(r"//", line)
```
Since this is going to parse a Lot of code and these same expressions are going to be hit a bunch, I'm curious if there might be some time benefit to pre-compiling the patterns with `re.compile(pat)` at the start of the program. From what I've read, the performance impact might be negligible but could also offer a nonzero speedup.

warriorstar-orion replied:
I'm compiling these now. The performance difference is probably negligible but this is cleaner regardless.

lewcc commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,337 @@
+
+def check_common_spelling_mistakes(filename, lines):
+    for line in lines:
+        if "nanotransen" in line.lower():
```
This feels like it could be made a bit more robust, but it's probably fine for this PR.

warriorstar-orion replied:
Ye

lewcc commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,337 @@
+def check_area_varedits(filename, lines):
+    for line in lines:
+        if re.match(r"^/area/.+[{]", line):
+            return "ERROR: Variable editted /area path use detected in a map, please replace with a proper area path."
```
```suggestion
            return "ERROR: Variable edited /area path use detected in a map, please replace with a proper area path."
```

warriorstar-orion replied:
Pulled the map checks but this will be in the following PR.

warriorstar-orion commented:
Temporarily removed checks that involve newlines due to the fact they won't catch issues if we're iterating over files line by line. I'll come up with a sane solution for those, but in the meantime, having the rest of these checks merged in would be beneficial.

warriorstar-orion commented:
I've added the tests involving newlines back, and fixed issues involving atoms with custom variables, so they're more reliable. In doing so I exposed about 2 dozen mapping issues that weren't being caught originally.

![2023_02_13__22_38_27__Port check_grep sh from an increasingly complicated shell script to Python  by w](https://user-images.githubusercontent.com/59303604/218633145-e2d4a70b-0b5a-4c9d-82df-0cf711d2b899.png)

So I have to fix these now.

Sirryan2002 commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,519 @@
+            return "ERROR: Unexpected commented out line detected in this map file. Please remove it."
+
+
+def check_conflict_markers(filename, lines):
+    for line in lines:
+        if "Map Merge Marker" in line:
+            return "ERROR: Merge conflict markers detected in map, please resolve all merge failures!"
```
any reason this function takes `filename` as a param but doesn't use it?

warriorstar-orion replied:
There was a dumb design issue where one of the checks required the filename in order to determine if it should be run on that file, the 515 syntax proc check. I've pulled this out as a separate check instead of trying to genericize everything for this, which you see in the unused filename arguments.

Sirryan2002 commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,519 @@
+            return "ERROR: Merge conflict markers detected in map, please resolve all merge failures!"
+
+
+def check_conflict_marker_object(filename, lines):
```
same for this function and all others below it

warriorstar-orion replied:
As above.

Sirryan2002 commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,519 @@
+def check_step_varedits(filename, lines):
+    for line in lines:
+        if re.match(r"step_[xy]", line):
+            return (
+                "ERROR: step_x/step_y variables detected in maps, please remove them."
+            )
```
```suggestion
            return "ERROR: step_x/step_y variables detected in maps, please remove them."
```
is there a specific reasons why this is formatted differently than the other return statements?

warriorstar-orion replied:
I was using black for formatting but I don't like how it handles these. I won't run a preformatter for now and maybe I can land on a coding style for this and the other Python code in the repo.

Edan52 commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,519 @@
+            for map_filepath in Path(".").glob("**/*.dmm"):
+                if map_filepath not in MAP_FILE_CACHE:
+                    MAP_FILE_CACHE[map_filepath] = open(map_filepath).read()
+                map = MAP_FILE_CACHE[map_filepath]
```
Might wanna use a different variable name

warriorstar-orion replied:
As above

Edan52 commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,519 @@
+def check_trailing_newlines(filename, lines):
+    last_line = [x for x in lines][-1]
+    if not last_line.endswith("\n"):
+        return f"ERROR: File {filename} is missing a trailing newline"
```
Can probably remove this and the filename parameters from the functions since it logs anyway

warriorstar-orion replied:
The logging here is super cleaned up and much more consistent in terms of return values for each check, as well as output format, which is now always "filename:lineno: message".

Edan52 commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,519 @@
+    failure = False
+
+    start = time.time()
+    for section, parts in BASIC_MAP_ISSUES.items():
```
For the map ones (if performance matters much) you'll probably be faster opening the files first and then running the checks on each one, though I appreciate that'll also cause some trouble with how it logs each test in order.

warriorstar-orion replied:
All the map checks are out of here now, happily

warriorstar-orion commented:
Thanks for the review @Edan52, all smart changes. Just responding so you know I'm not ignoring them :-) I'm still working through all of the individual cleanup tasks, but will come back to update the PR eventually.

warriorstar-orion commented:
As the latest commit message states, I've decided to break from the original formatting in order to allow for better performance by checking by file->check, instead of check->file, so each file only has to be opened once. The output format is similar to the DM compiler itself, being `filename:lineno: message`.

I've also pulled the map checks, which will arrive in a later PR, as the timeline for fixing up maps is much longer, and I don't want new whitespace issues to creep into the code while the PR is in review.

warriorstar-orion commented on tools/ci/check_grep2.py:
```diff
@@ -0,0 +1,519 @@
+
+            for map_filepath in Path(".").glob("**/*.dmm"):
+                if map_filepath not in MAP_FILE_CACHE:
+                    MAP_FILE_CACHE[map_filepath] = open(map_filepath).read()
```
All map checks have been removed from this PR, happily.
